### PR TITLE
Python Graph API / Strongly typed ID

### PR DESCRIPTION
**Pull request by xinyuli1204**
1. fix duplicate param name error when combine two graphs with same input name.
2. c_api for graph 
In short words,
argument_name now becomes  `std::string s = "_" + node_id + "_" + name + std::to_string(index) +  "_" + graph_id;`

### CODE REVIEW

**Comment by iitaku** [comment]
@xinyuli1204 Basically looks good. 

Now we have three different id (Node, Port and now Graph is added) and it has same type (std::string) even if these are not compatible each other. Can you think about introducing strongly typed alias for each id? ref: https://stackoverflow.com/questions/34287842/strongly-typed-using-and-typedef

File: include/ion/c_ion.h [hunk 1]
```diff
@@ -76,6 +78,12 @@ int ion_buffer_destroy(ion_buffer_t);
 int ion_buffer_write(ion_buffer_t, void *, int size);
 int ion_buffer_read(ion_buffer_t, void *, int size);
 
+int ion_graph_create(ion_graph_t *, ion_builder_t, const char *);
+int ion_graph_add_node(ion_graph_t, const char*, ion_node_t *);
+int ion_graph_destroy(ion_graph_t);
+int ion_graph_run(ion_graph_t);
+int ion_graph_create_with_multiple(ion_graph_t *, ion_graph_t obj1, ion_graph_t obj2);
```
Comment (iitaku) [inline include/ion/c_ion.h:85]: I'm considering more straitforward naming. e.g. `ion_graph_create_from_2` ? Do you have any better idea?

File: include/ion/node.h [hunk 2]
```diff
@@ -126,6 +127,31 @@ class Node {
...
     }
 
+    Node(const std::string& id, const std::string& name, const Halide::Target& target, const std::string& graph_id)
+        : impl_(new Impl{id, name, target, graph_id})
+    {
+    }
+
+
+    template<typename... Args>
+    std::vector<Port> create_ports_vector(Args... args) const {
```
Comment (iitaku) [inline include/ion/node.h:137]: Can you please clarify necessity of the wrapper function?

**Comment by xinyuli1204** [comment]
@iitaku  
My thought is:

## if no wrapper
 ### **Following example shows the input is the halide buffer** 
```
Buffer<int32_t> in0(size, size);
in0.fill(1);
Graph g0 = b.add_graph("graph0");
Node n0 = g0.add("test_inc_i32x2")(in0).set_param(Param("v", 40));  
n0["output"].bind(out0);
g0.run();
```

this code will call  constructor Port(const Halide::Buffer<T>& buf) and  bind buf to this port. `this->bind(buf);`
At this is time the graph_id is **empty,**  but later in `Node::set_iport `will pass the real graph_id.  this is not compatible and will throw error  `Argument "__b00_754500d7_38bf_473e_a506_7464433d99d7" is not found in the inferred arguments
 `in later `generate_arguments_instance `function`. arg.name is "__b00_" (since graph-id has not passed when Port is ocnstructed)
 
(argument is 
`previous node id  +  previous name + current input port name + current graph_id`)
```
 template<typename... Args>
    Node operator()(Args ...args) {
        set_iport(std::vector<Port>{args...});
        return *this;
    }
```
```
void Node::set_iport(Port port) {
    port.impl_ ->graph_id = impl_->graph_id;
    port.impl_->succ_chans.insert({id(), port.pred_name()});
    impl_->ports.push_back(port);
}
```


So I want to pass the real graph_id when in port constructor  if input is halide buffer and an array of halide buffer

 ```
template<typename... Args>
    Node operator()(Args ...args) {
        set_iport(create_ports_vector(args...));
        return *this;
    }

    Port get_iport(Port arg) const {
        // if input is port
        return arg;
    }

   template<typename T>
   Port get_iport(Halide::Buffer<T>&  arg) const {
        // if input is halide buffer
        if (impl_->graph_id.empty())
            return Port(arg);
        else
            return Port(arg, impl_->graph_id);
   }

    template<typename... Args>
    std::vector<Port> create_ports_vector(Args... args) const {
        return {get_iport(args)...};
    }

   template<typename T>
   Port get_iport(std::vector<Halide::Buffer<T>>& arg) const {
        // if input is halide buffer
        if (impl_->graph_id.empty())
            return Port(arg);
        else
            return Port(arg, impl_->graph_id);
   }


```


 ### **this example shows the input is port** 
```
 Port ip0{"input", Halide::type_of<int32_t>(), 2}; //graph -id is empty
 Port vp0{"v", Halide::type_of<int32_t>()}; //graph -id is empty

Graph g0 = b.add_graph("graph0");  //generating graph -id

Node n0 = g0.add("test_incx_i32x2")(ip0,vp0);

ip0.bind(ibuf0);  //graph -id is passed to port
int32_t v0 = 0;
vp0.bind(&v0);   //graph -id is passed to port
```
there is no error when still  use 
```
 template<typename... Args>
    Node operator()(Args ...args) {
        set_iport(std::vector<Port>{args...});
        return *this;
    }
```
because the port is still constructed with empty graph id but will bind the buffer later when graph_id is not empty.

so both   impl_->params and as_argument() has non-empty grapth-id 
`generate_arguments_instance`  won't throw error.

File: include/ion/node.h (hunk 2 above)
Comment (xinyuli1204) [reply include/ion/node.h]: Sorry for the confusion,  this is unnecessary and I delete it

**Comment by xinyuli1204** [comment]
> @xinyuli1204 Basically looks good.
> 
> Now we have three different id (Node, Port and now Graph is added) and it has same type (std::string) even if these are not compatible each other. Can you think about introducing strongly typed alias for each id? ref: https://stackoverflow.com/questions/34287842/strongly-typed-using-and-typedef

@iitaku  done
using NodeID = StringID<node_tag>;
using GraphID = StringID<graph_tag>;
using PortID = StringID<port_tag>;

File: include/ion/node.h [hunk 3]
```diff
@@ -126,6 +131,40 @@ class Node {
     {
     }
 
+    Node(const std::string& id, const std::string& name, const Halide::Target& target, const GraphID& graph_id)
+        : impl_(new Impl{id, name, target, graph_id})
+    {
+    }
+
+    Port get_iport(Port arg) const {
```
Comment (iitaku) [inline include/ion/node.h:139]: Probably `make_iport` is less confusing for this internal API.

File: include/ion/port.h [hunk 4]
```diff
@@ -78,58 +79,87 @@ class Port {
      * @arg k: The key of the port which should be matched with BuildingBlock Input/Output name.
      * @arg t: The type of the value.
      */
-    Port(const std::string& n, Halide::Type t) : impl_(new Impl("", n, t, 0)), index_(-1) {}
+    Port(const std::string& n, Halide::Type t) : impl_(new Impl("", n, t, 0,  GraphID(""))), index_(-1) {}
```
Comment (iitaku) [inline include/ion/port.h:82]: Is `GraphID("")` and `GraphID()` identical? Let's use either one if there is no meaning for these two use

File: include/ion/port.h [hunk 5]
```diff
@@ -220,7 +250,7 @@ class Port {
                  args.push_back(Halide::Var::implicit(i));
                  args_expr.push_back(Halide::Var::implicit(i));
              }
-             Halide::Func f(param.type(), param.dimensions(), argument_name(pred_id(), pred_name(), i) + "_im");
+             Halide::Func f(param.type(), param.dimensions(), argument_name(pred_id(), pred_name(), i, graph_id_to_string()) + "_im");
```
Comment (iitaku) [inline include/ion/port.h:253]: We can pass GraphID to `argument_name` directly

File: include/ion/port_map.h [hunk 6]
```diff
@@ -16,7 +16,7 @@ class PortMap {
     template<typename T>
     [[deprecated("Port::bind can be used instead of PortMap.")]]
     void set(Port port, T v) {
-        auto& buf(scalar_buffer_[argument_name(port.pred_id(), port.pred_name(), port.index())]);
+        auto& buf(scalar_buffer_[argument_name(port.pred_id_to_string(), port.pred_name(), port.index(), port.graph_id_to_string())]);
```
Comment (iitaku) [inline include/ion/port_map.h:19]: argument name takes NodeID and GraphID, not string

**Comment by xinyuli1204** [comment]
@iitaku can I ask something about pred_chan in port.h? 
From my understanding pred_chan makes tuple of NodeId and string  `std::tuple<NodeID, std::string>;`, but when constructing new port  in port.cc 

```
Port::Impl::Impl(const std::string& pid, const std::string& pn, const Halide::Type& t, int32_t d, const GraphID & gid)
    : id(sole::uuid4().str()), pred_chan{pid, pn}, succ_chans{}, type(t), dimensions(d), graph_id(gid)
```

**pred_chan{pid, pn},** why it takes pid instead?

**Comment by Fixstars-iizuka** [comment]
@xinyuli1204 It's correct. `pid` is actually `nid`. 

PortID is not used at this moment, but if it has no down effects, let's leave it as it is for the future.

**Pull request merged**